File: 783. Minimum Distance Between BST Nodes/dfs.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int res = 100001;
    TreeNode* prev;
    int minDiffInBST(TreeNode* root) {
        if (root == NULL) return 0;
        helper(root);
        return res;
    }

    void helper(TreeNode* root) {
        if (root == NULL) return;
        helper(root->left);
        if (prev != NULL) {
            res = min(res, root->val - prev->val);
        }
        prev = root;
        helper(root->right);
    }
};

File: 783. Minimum Distance Between BST Nodes/dfs.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    res = 100001
    prev = None
    
    def minDiffInBST(self, root: Optional[TreeNode]) -> int:
        if root is None:
            return 0
        
        self.helper(root)
        
        return self.res
    
    def helper(self, root):
        if root is None:
            return
        
        self.helper(root.left)
        
        if self.prev is not None:
            self.res = min(self.res, root.val - self.prev)
        
        self.prev = root.val
        
        self.helper(root.right)

        
        
